### PR TITLE
TextureAny::from_id

### PR DESCRIPTION
**Pull request by ryanstew**
Add from_id to texture types to allow them to point to an externally-allocated texture. See #1072 for details.

### CODE REVIEW

File: src/texture/any.rs [hunk 1]
```diff
@@ -69,6 +69,53 @@ pub struct TextureAny {
...
+        Dimensions::Texture2dArray { .. } => gl::TEXTURE_2D_ARRAY,
+        Dimensions::Texture2dMultisample { .. } => gl::TEXTURE_2D_MULTISAMPLE,
+        Dimensions::Texture2dMultisampleArray { .. } => gl::TEXTURE_2D_MULTISAMPLE_ARRAY,
+        Dimensions::Texture3d { .. } => gl::TEXTURE_3D,
+        Dimensions::Cubemap { .. } => gl::TEXTURE_CUBE_MAP,
+        Dimensions::CubemapArray { .. } => gl::TEXTURE_CUBE_MAP_ARRAY,
+    }
+}
+
+unsafe fn generate_mipmaps(ctxt: &CommandContext,
```
Comment (ryanstew) [inline src/texture/any.rs]: Open question: where mipmaps are regenerated on upload, this logic is slightly different (the GLES check is not in place). Is that a bug, or intentional?

File: tests/texture_draw.rs [hunk 2]
```diff
@@ -34,30 +30,81 @@ macro_rules! texture_draw_test {
...
+            let (vb, ib) = support::build_rectangle_vb_ib(&display);
+
+            let program = create_program!(display, $glsl_ty, $glsl_value);
+
+            let empty_texture = glium::texture::$tex_ty::empty_with_format(&display,
+                                                                           glium::texture::$format_ty::$format_value,
+                                                                           glium::texture::MipmapsOption::NoMipmap,
+                                                                           $($dims),+).unwrap();
+            let texture = unsafe {
+                glium::texture::$tex_ty::from_id(&display,
```
Comment (ryanstew) [inline tests/texture_draw.rs]: I assume it's safe to just repeat the format here and trust that Glium didn't change the underlying format, as I can't find an easy way to convert the internal format of a texture back out into an UncompressedFloatFormat.

File: src/texture/any.rs (hunk 1 above)
Comment (tomaka) [inline src/texture/any.rs]: That would be a bug.

**Comment by matklad** [comment]
May I suggest changing the type of `owned: bool` from `bool` to two element `enum` in the sake of readability? 

At the call site, `Owner::Callee` would be more self documenting then `true`.

**Comment by tomaka** [comment]
Let's merge this.

The code is not very clean, but the textures code in general is really dirty.

**Pull request merged**